RecoTauTag/RecoTau/plugins/DeepTauId.cc

```diff
@@ -265,20 +590,52 @@ class DeepTauId : public deep_tau::DeepTauBase {
         desc.add<edm::ParameterSetDescription>("VSeWP", descWP);
         desc.add<edm::ParameterSetDescription>("VSmuWP", descWP);
         desc.add<edm::ParameterSetDescription>("VSjetWP", descWP);
-        descriptions.add("DeepTau2017v1", desc);
+        descriptions.add("DeepTau", desc);
     }
 
 public:
     explicit DeepTauId(const edm::ParameterSet& cfg, const deep_tau::DeepTauCache* cache) :
         DeepTauBase(cfg, GetOutputs(), cache),
-        electrons_token(consumes<ElectronCollection>(cfg.getParameter<edm::InputTag>("electrons"))),
-        muons_token(consumes<MuonCollection>(cfg.getParameter<edm::InputTag>("muons"))),
-        input_layer(cache_->getGraph().node(0).name()),
-        output_layer(cache_->getGraph().node(cache_->getGraph().node_size() - 1).name())
+        electrons_token_(consumes<ElectronCollection>(cfg.getParameter<edm::InputTag>("electrons"))),
+        muons_token_(consumes<MuonCollection>(cfg.getParameter<edm::InputTag>("muons"))),
+        rho_token_(consumes<double>(cfg.getParameter<edm::InputTag>("rho"))),
+        version(cfg.getParameter<unsigned>("version")),
+        debug_level(cfg.getParameter<int>("debug_level"))
     {
-        const auto& shape = cache_->getGraph().node(0).attr().at("shape").shape();
-        if(shape.dim(1).size() != dnn_inputs_2017v1::NumberOfInputs)
-            throw cms::Exception("DeepTauId") << "number of inputs does not match the expected inputs for the given version";
+        if(version == 1) {
+            input_layer_ = cache_->getGraph().node(0).name();
+            output_layer_ = cache_->getGraph().node(cache_->getGraph().node_size() - 1).name();
+            const auto& shape = cache_->getGraph().node(0).attr().at("shape").shape();
+            if(shape.dim(1).size() != dnn_inputs_2017v1::NumberOfInputs)
+                throw cms::Exception("DeepTauId") << "number of inputs does not match the expected inputs for the given version";
+        } else if(version == 2) {
+            tauBlockTensor_ =  std::make_shared<tensorflow::Tensor>(tensorflow::DT_FLOAT, tensorflow::TensorShape{ 1,
```

Comment:
I'm curious if you tried float16.
This should reduce the memory size and will also speed up the vectorizable part.

Reply:
@slava77 Interesting idea, thanks. The type of inputs should be the same as the weights of the network, otherwise, it gives errors similar to that one:
```Input 'y' of 'Mul' Op has type float16 that does not match type float32 of argument 'x'```
So it would require to convert the whole graph to use float16. There is no automated tool within tensorflow to convert graph from float32 to float16. Another way around could be to recreate NN model in Keras with float16 and set weights manually.  While trying this, I got an error that is related to an open issue related to the support of float16 in BatchNormalization: https://github.com/keras-team/keras/issues/9582
I'll try to see if there are some other ways to do it, but it could take some time.